### ISSUE TITLE
Actix-web 2.0 is leaking memory in every request

### ISSUE THREAD

**Issue by Lesiuk** [actix/actix-web#1446]
Your issue may already be reported!
Please search on the [Actix Web issue tracker](https://github.com/actix/actix-web/issues) before creating one.

## Steps to Reproduce (for bugs)

1.
2.
3.
4.

## Your Environment

* Rust Version (I.e, output of `rustc -V`):
* Actix Web Version:

**Comment by mehmetsefabalik** [comment]
Which example repo did you compile?

**Comment by Lesiuk** [comment]
Recorded little demonstration of the issue on youtube https://www.youtube.com/watch?v=nQ0YxJ94B2s 

It's example from Readme

**Comment by robjtede** [comment]
I've got circumstancial evidence (production service recently updated) of there being a leak introduced somewhere in v2.

**Comment by cdbattags** [comment]
@robjtede Do we have any leads as to where?

**Comment by robjtede** [comment]
@cdbattags Not yet, personally noticed only recently, but we should keep an eye out and post any discoveries here.

**Comment by Lesiuk** [comment]
@cdbattags I was able to debug it with MTuner and it showed me where are those allocations, but after turning it off It doesn't want to turn on again.

![image](https://user-images.githubusercontent.com/543264/78907538-1e8abc80-7a81-11ea-9a90-e32e6b2aa7e6.png)

**Comment by cdbattags** [comment]
Ok no problem @Lesiuk, sooner the better on getting any sort of debug information. This should be decently easy to squash we just need more data.

**Comment by Lesiuk** [comment]
![image](https://user-images.githubusercontent.com/543264/78910258-f604c180-7a84-11ea-89bf-7661e9c75da8.png)

**Comment by Lesiuk** [comment]
First actix line is https://github.com/actix/actix-net/blob/master/actix-rt/src/arbiter.rs#L185

**Comment by Lesiuk** [comment]
tokio::runtime:task::core::Cell is allocating memory which is never deallocated.

**Comment by Lesiuk** [comment]
This change looks suspicious, I will build it with older tokio https://github.com/tokio-rs/tokio/commit/a78b1c65ccfb9692ca5d3ed8ddde934f40091d83

**Comment by cdbattags** [comment]
Which example are you using? Can you give minimal repro steps?

**Comment by cdbattags** [comment]
@Lesiuk why do you think that commit looks suspicious?

**Comment by cdbattags** [comment]
Looks like first usage of `unsafe` in this chain is https://github.com/tokio-rs/tokio/blob/master/tokio/src/task/local.rs#L200

I know that's not much to go off of but hmmm

**Comment by cdbattags** [comment]
@carllerche or @alexcrichton any ideas?

**Comment by Lesiuk** [comment]
![image](https://user-images.githubusercontent.com/543264/78914430-aaedad00-7a8a-11ea-8349-92723f46c45c.png)

Is fixing the issue 
![image](https://user-images.githubusercontent.com/543264/78914467-b2ad5180-7a8a-11ea-9c99-02b5b8b6b1f8.png)

**Comment by cdbattags** [comment]
Ok, so then please let's promote this as a proper tokio issue!

**Comment by cdbattags** [comment]
@Lesiuk would you like to create it or shall I?

FYI definitely a Tokio issue @alexcrichton and @carllerche

**Comment by robjtede** [comment]
@Lesiuk deleted your screencap comment because?

**Comment by Lesiuk** [comment]
Because I'm still debugging and I think I found out something.

It's working fine with actix-rt 1.0 and leaking with actix-rt 1.1.

**Comment by Lesiuk** [comment]
Let's see what were the changes between actix-rt 1.1 and 1.0.

**Comment by Lesiuk** [comment]
There is a change in the line where it's leaking

![image](https://user-images.githubusercontent.com/543264/78916582-fa81a800-7a8d-11ea-9a44-7fe7c2e7fb98.png)

**Comment by robjtede** [comment]
https://github.com/actix/actix-net/commit/f1685d82534e41a36ccff783e6af631d81db5247#diff-01ce174553e287daa3a0debf474e11c7

L347

**Comment by Lesiuk** [comment]
Aren't elements never removed from this PENDING array?

**Comment by robjtede** [comment]
@Jonathas-Conceicao can you shed some light on this change?

**Comment by Jonathas-Conceicao** [comment]
The source of the increasing memory usage is the indeed the PENDING array. A call for `actix_rt::Arbiter::local_join` would replace the PENDING vector allowing memory to be freed, this small example can test it interactively.

```Rust
use actix_web::{get, post, App, HttpServer};

#[get("/")]
async fn hello() -> String {
    format!("Hello")
}

#[post("/")]
async fn free() -> String {
    let _ = actix_rt::Arbiter::local_join();
    format!("Clear")
}

#[actix_rt::main]
async fn main() -> std::io::Result<()> {
    HttpServer::new(|| App::new().service(hello).service(free))
        .bind("127.0.0.1:8080")?
        .run()
        .await
}
```

Nevertheless I do admit it's a poor design on my part for https://github.com/actix/actix-net/pull/94 and I'll try to think of a proper solution for it.

**Comment by cdbattags** [comment]
Sorry for my hastiness. I just want to make sure we're being proactive on these types of issues because we're getting a lot of bad "PR" for this 2+ release as it stands and a memory leak is the worst kind of news.

What's the ETA on a fix?

**Comment by Jonathas-Conceicao** [comment]
Since the leak only happens on `actix-rt:1.1.0`, users can just enforce a more stable version with `actix-rt = "=1.0.0"` on their `Cargo.toml`. As for the fix I'll have some time to work on it later today.

**Comment by Lesiuk** [comment]
Maybe we can yank actix-rt:1.1.0 ?

**Comment by robjtede** [comment]
I agree yank it. Any point releasing the currect version as `1.2.0-beta.1` for testing reasons?

ping @JohnTitor, thoughts on yanking?
